### PR TITLE
Fix/fix camera bounds

### LINKED ISSUE #80: Incorrect camera bounds
## Description
- Expected behavior: Camera should be centered at locations such that the entire screen is filled with the background image
- Actual behavior: On some devices, camera bounds is not calculated correctly, resulting in black spaces that are not covered by the background image

iPad Pro Simulator:

iPad Pro

iPad (10th generation) Simulator:

Simulator Screenshot - iPad (10th generation) - 2024-04-07 at 15 07 29

## Potential Causes
- The screen size is fetched from `UIScreen.main.bounds.size` in `GameScene.swift` and `SDCameraObject` calculates the bounds using this screen size
- Apparently, `UIScreen.main.bounds.size` can returns screen size using launch image? But currently, we do not have launch images
- https://stackoverflow.com/questions/21668497/uiscreen-mainscreen-bounds-returning-wrong-size

## Potential Solutions
- `GeometryReader`

### PR DESCRIPTION
**Pull request by jasonqiu212**
Resolves #80 

## Solution Description
- By setting our game scene's `SKSceneScaleMode = .aspectFill`, SpriteKit scales the scene's area by applying the larger of the both axis's scaling factor.
    - E.g. Scene size: 5042x1040 and device size: 1180x820
    - X-axis scaling factor: 1180/5042=0.234
    - Y-axis scaling factor: 820/1040=0.719
    - Thus, the scene's x and y-axis are both scaled by 0.719 to fit into the view
- Since the scene is scaled, the boundaries of the scene must depend on the **scaled** device size, rather than the raw device size

## Note
- Cameras apply a scale that scales the map to the view height of 1040pts, to prevent the player from looking small in huge maps.
- Refactored the checking of camera bounds to use `SKConstraint`

### CODE REVIEW

File: SDPhysicsEngine/Sources/SDPhysicsEngine/GameScene.swift [hunk 1]
```diff
@@ -81,17 +74,26 @@ extension GameScene: SDScene {
         object.removeFromParent()
     }
 
-    private func playerScreenSize(for numberOfPlayers: Int) -> CGSize {
+    private func scaledPlayerScreenSize(for numberOfPlayers: Int) -> CGSize {
```
Comment (Junyi00) [inline SDPhysicsEngine/Sources/SDPhysicsEngine/GameScene.swift:77]: GameScene ideally should be blind to the layout of the screen, or at least the size of the rendered view should not be assumed here (it should be received from some source like the renderer). This leads to some hidden coupling between the renderer and the game scene.

File: SDPhysicsEngine/Sources/SDPhysicsEngine/Object/SDCameraObject.swift [hunk 2]
```diff
@@ -3,51 +3,40 @@ import SpriteKit
...
         self.cameraNode = SKCameraNode()
         self.player = player
 
-        let boundsOrigin = CGPoint(x: screenSize.width / 2, y: screenSize.height / 2)
-        let boundsSize = CGSize(width: sceneSize.width - screenSize.width, height: sceneSize.height - screenSize.height)
-        self.bounds = CGRect(origin: boundsOrigin, size: boundsSize)
-        if sceneSize.height > 1_024 {
-            scale = 1_024 / sceneSize.height
-        }
+        let scale = sceneSize.height > 1_200 ? 0.5 : 1
```
Comment (Junyi00) [inline SDPhysicsEngine/Sources/SDPhysicsEngine/Object/SDCameraObject.swift:17]: Where did we get the magic number of 1200 from?

**Review by Junyi00 — CHANGES_REQUESTED** [review]

File: SDPhysicsEngine/Sources/SDPhysicsEngine/GameScene.swift (hunk 1 above)
Comment (jasonqiu212) [reply SDPhysicsEngine/Sources/SDPhysicsEngine/GameScene.swift]: Mm I understand. I agree that the `Renderer` should be the one dealing with the rendered screen's size.

My changes currently refactor the flow of updating the camera constraints like this:
1. When `GameBridge` needs to create a new player object, it calls `GameScene::addPlayerObject`
2. `GameScene::addPlayerObject` creates a basic camera with the player, but without any of the constraints and scale
3. `GameScene::addPlayerObject` delegates setting up the camera constraints and scale to its `sceneDelegate` (i.e. `GameViewController::setupCameras`)
4. `GameViewController::setupCameras` calls `MTKRenderer::setupCameras`, which derives the player screen size from `LayoutUtils` and calls `GameScene::setupCameras`
5. `GameScene::setupCameras`, now with the player screen size, then calls `SDCameraObject::setup` to set the constraints and camera scale

Do you think this is too convoluted?

Comment (Junyi00) [reply SDPhysicsEngine/Sources/SDPhysicsEngine/GameScene.swift]: I think it feels a little convoluted, not too sure if the viewcontroller needs to be involved anymore... 
I think MTKRenderer can setup the constraints and scale in `updateScene` where it calls `scene.useCamera`, then `GameScene` just record if the constraints/scale has been changed to ensure it only does it once?

Comment (jasonqiu212) [reply SDPhysicsEngine/Sources/SDPhysicsEngine/GameScene.swift]: good idea!

**Pull request merged**